Refactor `tuftool` to use `RepositoryEditor`

zmrow opened the pull request:
*Issue #, if available:*
Fixes #43 

*Description of changes:*
The tl;dr is: "simplify `tuftool` by using `RepositoryEditor`."

The actual list of changes is:
* Add "getters" to `SignedRole` so a user can access, but not change data.
* Use `RepositoryEditor` anywhere a repository is being manipulated inside `tuftool`. This includes the `create` and `refresh` commands.
* **Breaking change**: Move the `sign` command to a subcommand of `tuftool root`. This functionality is primarily meant of creation of the `root.json` anyway.
* **Breaking change**: `tuftool create` now *symlinks* targets to the given directory rather than *copying* them. This is a better use of disk space as repositories or targets get larger. *If we still want to keep the copy functionality we can provide a flag the user can pass?*
* Remove any unused files
* Remove any unused error variants

*Testing done:*
* All unit and integration tests pass (which gives me pretty good confidence that we're creating repos the right way)
* Manually create a repository locally; targets are named correctly and properly symlinked.
* Manually refresh a repository locally; everything looks good here as well.


By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

zmrow committed:
Refactor tuftool to use RepositoryEditor

zmrow committed:
Add getters to SignedRole

This allows a user to read, but not modify the data inside
SignedRole.

tjkirch commented on tuftool/src/error.rs:
```diff
@@ -1,6 +1,8 @@
 // Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 // SPDX-License-Identifier: MIT OR Apache-2.0
 
+// Not really worried about the memory penalty of large enum variants here
+#![allow(clippy::large_enum_variant)]
```
What makes it large?  Nothing particularly stands out to me, and it seems smaller than it did before...

zmrow replied:
Clippy is angry that some of the enum variants are larger than others. I really think it's overly pedantic in this case and for a CLI tool I don't think we care about this.
```
warning: large size difference between variants
   --> tuftool/src/error.rs:145:5
    |
145 | /     LinkTarget {
146 | |         indir: PathBuf,
147 | |         outdir: PathBuf,
148 | |         source: tough::error::Error,
149 | |         backtrace: Backtrace,
150 | |     },
    | |_____^
```

tjkirch commented on tuftool/src/root.rs:
```diff
@@ -85,6 +87,13 @@ pub(crate) enum Command {
         #[structopt(short = "r", long = "role")]
         roles: Vec<RoleType>,
     },
+    /// Set the signature count threshold for a role
+    Sign {
```
I think the comment was copypasta :)

zmrow replied:
Oops

tjkirch commented on tuftool/src/root.rs:
```diff
@@ -262,6 +272,21 @@ impl Command {
+        )
+        .context(error::SignRoot { path })?;
+
+        std::fs::write(&path, signed_root.buffer()).context(error::FileWrite { path: &path })?;
```
Would be pretty funny to partially overwrite someone's root key with half of the signed data :D  Perhaps we should write to a temp file and move it over the original path?  (nit)

zmrow replied:
Good point. :) I'll fix that.

tjkirch commented on tuftool/src/create.rs:
```diff
@@ -77,202 +63,49 @@ pub(crate) struct CreateArgs {
+        signed_repo.write(metadata_dir).context(error::WriteRepo)?;
+        signed_repo
+            .link_targets(&self.indir, targets_dir)
+            .context(error::LinkTarget {
```
```suggestion
            .context(error::LinkTargets {
```

tjkirch commented on tuftool/src/error.rs:
```diff
@@ -253,6 +254,12 @@ pub(crate) enum Error {
     },
 
+    #[snafu(display("Failed writing repo data to disk: {}", source))]
+    WriteRepo {
```
Can we include the directory where we were trying to write the repo?

tjkirch commented on tuftool/src/create.rs:
```diff
@@ -77,202 +63,49 @@ pub(crate) struct CreateArgs {
 impl CreateArgs {
     pub(crate) fn run(&self) -> Result<()> {
-        if let Some(jobs) = self.jobs {
-            rayon::ThreadPoolBuilder::new()
```
rayon didn't seem to actually be used in refresh.rs, so I understand that removal, but I think it's used here in create to parallelize building targets, and I think we're losing that with the move to the repo editor... right?  Does it take a lot longer to create repos?  I wonder if we should include that as an improvement to repo editor in another commit for this PR?

tjkirch replied:
Per offline discussion, tough shouldn't own the parallelism, but tuftool could/should.  If we can add `Target`s individually to the editor, rather than just paths, then we can build `Target`s in parallel in tuftool and add them to the editor as they're ready.  I don't think it has to be done here, but it does feel unfortunate to lose the parallelism with the change...

webern replied:
We should probably add Windows to the build matrix. Easy to do, seems relevant with respect to symlinks, unless they 'just work' the same over there.

tjkirch commented on tuftool/src/create.rs:
```diff
@@ -77,202 +63,49 @@ pub(crate) struct CreateArgs {
+
+        let metadata_dir = &self.outdir.join("metadata");
+        let targets_dir = &self.outdir.join("targets");
+        signed_repo.write(metadata_dir).context(error::WriteRepo)?;
+        signed_repo
```
Should we link targets before we write metadata so the metadata is valid/usable as soon as it's there?

zmrow replied:
I'll change it as you're right, but I can't see there being a race condition of any kind here unless a user is using their local machine as a "real" repo and polling for updates constantly. That being said... we have no idea what a user is going to do. :)

tjkirch commented on tuftool/src/create.rs:
```diff
@@ -77,202 +63,49 @@ pub(crate) struct CreateArgs {
-                    length += n as u64;
-                }
+    // Build a list of files from the given directory
+    fn build_targets(&self) -> Result<Vec<PathBuf>> {
```
```suggestion
    fn find_targets(&self) -> Result<Vec<PathBuf>> {
```
Not really building anymore...

tjkirch reviewed (commented):
Look at all that beautiful red!

zmrow commented:
Fixes @tjkirch 's comments except for the multithreading question and the clippy comment.

zmrow commented:
@tjkirch @iliana : This PR symlinks the targets to the given outdir rather than copying them. How do you feel about that? Should we provide a flag so the user can choose to copy the targets?

tjkirch commented:
> @tjkirch @iliana : This PR symlinks the targets to the given outdir rather than copying them. How do you feel about that? Should we provide a flag so the user can choose to copy the targets?

You know I like it, I requested it :)  A flag would be nice, I guess, but I'm not sure of a reason for it yet.

zmrow commented:
> > @tjkirch @iliana : This PR symlinks the targets to the given outdir rather than copying them. How do you feel about that? Should we provide a flag so the user can choose to copy the targets?
> 
> You know I like it, I requested it :) A flag would be nice, I guess, but I'm not sure of a reason for it yet.

Unless @iliana has different feelings, I'll leave it out for now.

webern reviewed (approved):
Wow, now I see why the editor is a good thing!

zmrow committed:
Update targets methods in RepositoryEditor

This commit changes the `add_target` and `add_targets` methods
to `add_target_path` and `add_target_paths`. This is more descriptive
of their purpose. It also adds an `add_target` method, which takes
in a filename and a Target struct, which allows a user to parallelize
the creation of their targets.

zmrow commented:
The above commit adds the threading support back into `tuftool create`. I ended up being able to use a decent amount of the original code, so I got that going for me too, which is nice.

zmrow commented:
Rebased onto `develop`

tjkirch commented on tough/src/editor/mod.rs:
```diff
@@ -189,8 +189,21 @@ impl RepositoryEditor {
     }
 
+    /// Add a `Target` to the repository
+    pub fn add_target(&mut self, name: String, target: Target) -> Result<&mut Self> {
```
It won't let me comment below, but we should be able to use this at the end of `add_target_path` now.

zmrow replied:
ha good catch - I'll fix it

tjkirch replied:
That change made me realize that this function doesn't have to return a result - we can always insert into the hash, and so we can remove the burden of error checking from callers.

zmrow replied:
You're right... tunnel vision over here. I'll fix that before I merge.

tjkirch commented on tough/src/editor/test.rs:
```diff
@@ -58,19 +58,23 @@ mod tests {
         .unwrap();
-        let target3 = targets_path().join("file3.txt");
+        let target3_path = targets_path().join("file3.txt");
+        let target2_path = targets_path().join("file2.txt");
```
Hmm, should we name this something other than what's in the existing targets?  Or maybe test both, replacing an existing target and adding a new one from a `Target`?

zmrow replied:
I'll rename this to something else like "file4.txt"

tjkirch commented on tuftool/src/refresh.rs:
```diff
@@ -75,124 +68,43 @@ pub(crate) struct RefreshArgs {
 
 impl RefreshArgs {
     pub(crate) fn run(&self) -> Result<()> {
-        if let Some(jobs) = self.jobs {
```
It won't let me comment on the new location of this, but it made me think - the comment on `jobs` says it defaults to the number of cores, but there's no structopt-level default, and here we're checking if the argument was given... are we actually getting parallelism?  Do we need a default in the args struct?

zmrow replied:
Good question. Per the [`rayon` FAQ](https://github.com/rayon-rs/rayon/blob/master/FAQ.md), it will spawn the same number of threads as cores. It does this by default when you use any of the parallel iterator methods it provides, which we do below in `build_targets()`.

If a user specifies a specific number of jobs, we use `rayon::ThreadPoolBuilder` to override the default.

tjkirch replied:
Ah, sorry, I was silly and forgot that rayon creates a global thread pool by default.  This is just configuring it, if the user wanted a different number of threads.

zmrow replied:
I'll add a comment for future readers.

zmrow commented:
Addressed @tjkirch 's comments

zmrow commented:
Removed the `Result<>` from the `add_target()` method. Good catch @tjkirch !

zmrow commented:
Fixed the clippy issue.
